# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Data Sets

In [2]:
# defining path
path=(r'C:\Users\Admin\Desktop\Data Tasks\Data Immersion\Achievement 4\Instacart Basket Analysis\02 data\Original Data')

In [3]:
# importing dataset orders
df_ords=pd.read_csv(os.path.join(path,'orders.csv'), index_col=False)

In [4]:
# importing dataset departments
df_dep=pd.read_csv(os.path.join(path,'departments.csv'),index_col=False)

In [5]:
# importing dataset products
df_prods=pd.read_csv(os.path.join(path,'products.csv'),index_col=False)

# Direction 1: Data Wrangling (data frame orders )

#### Original data frame overview

In [6]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Dropping Columns

#### 1. We will drop the column eval_set since we don't need it for analysis

In [7]:
# dropping column 'eval_set'
df_ords=df_ords.drop(columns=['eval_set'])

#### outcome

In [8]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### 2. The days_since_prior_order column has empty values(NaN), we will take a closer look into it, if there is too much missing data in that column we can drop it too.

In [9]:
# Checkin the NaN values in the days_since_prior_order column
df_ords['days_since_prior_order'].value_counts(dropna=False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

#### Result: The total number of empty cells is 206209, this makes approximately 6 % of the total 3421083 rows. In this case we don't drop the column because the missing data does not exceed the 10% of the total data.

#

## Renaming Columns

In [10]:
# renaming the 'order_dow' column name with a more intuitive one
df_ords.rename(columns={'order_dow':'orders_day_of_week'},inplace=True)

#### outcome

In [11]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Changing a Variable’s Data Type

#### We will change the data type of the column 'order_id', because we don't want it to be perceived as a numerical data and be included in descriptive statistics

In [12]:
# changing data type for column 'order_id'
df_ords['order_id']=df_ords['order_id'].astype('str')

#### otucome

In [13]:
df_ords['order_id'].dtype

dtype('O')

## Transposing Data ( data frame departments)

#### Original data frame overview

In [14]:
df_dep

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

#### Transposing df_dep

In [15]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [16]:
# overwriting the df_dep
df_dep_t=df_dep.T

#### outcome

In [17]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
# indexing row 0 as the new header
new_header = df_dep_t.iloc[0]

#### outcome

In [19]:
new_header

0    department
Name: department_id, dtype: object

In [20]:
# copying all the rows except 0 (new_header)
df_dep_t_new = df_dep_t[1:]

In [21]:
# creating new df without row 0
df_dep_t_new = df_dep_t[1:]

In [22]:
# adding the headers to the new df
df_dep_t_new.columns=new_header

#### outcome

In [23]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#### voilà

In [49]:
df_dep_t_new[df_dep_t_new.department.isin(['breakfast','deli'])]

department_id department
14             breakfast
20                  deli

# Direction 2:

#### Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [24]:
# changing dtype of 'user_id'
df_ords['user_id']=df_ords['user_id'].astype('str')

#### outcome

In [25]:
df_ords['user_id'].dtypes

dtype('O')

# Direction 3:

#### Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [26]:
# checking column names
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [27]:
# viewing the 'order_hour_of_day' column name as 'order_time'
df_ords.rename(columns={'order_hour_of_day':'order_time'})

order_id user_id  order_number  orders_day_of_week  order_time  \
0        2539329       1             1                   2           8   
1        2398795       1             2                   3           7   
2         473747       1             3                   3          12   
3        2254736       1             4                   4           7   
4         431534       1             5                   4          15   
...          ...     ...           ...                 ...         ...   
3421078  2266710  206209            10                   5          18   
3421079  1854736  206209            11                   4          10   
3421080   626363  206209            12                   1          12   
3421081  2977660  206209            13                   1          12   
3421082   272231  206209            14                   6          14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

#### Note: the name hasn't been overwritten

In [28]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

# Direction 4:

#### Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [29]:
# finding the frequency for column 'order_hour_of_day'
df_ords['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#### Answer: the most orders have been made from 10:00 to 16:00, with the busiest hour being 10 am 

# Direction 5:

#### Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [30]:
# transforming 'departments' data frame into a data dictionary
data_dict=df_dep_t_new.to_dict('index')

In [31]:
#getting the department name with id 4
print(data_dict.get('4'))

{'department': 'produce'}


#### Answer: the department name for the 'department_id' with the value 4 is 'produce'

# Direction 6:

#### The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [32]:
#finding the departmet_id for 'breakfast'
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [33]:
# creating the subset for breakfast department
df_breakfast=df_prods[df_prods['department_id']==14]

#### outcome

In [34]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# Direction 7:

#### They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [35]:
#finding the departmet_id for 'breakfast'
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [36]:
# creating the subset(dasaframe dinner parties) for departments:alcohol, deli, beverages, and meat/seafood
df_din_prt=df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

#### outcome

In [37]:
df_din_prt

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Direction 8:

#### It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [44]:
df_din_prt.shape

(7650, 5)

#### answer: in the dinner party subset dataframe has 7650 rows

# Direction 9:

#### Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [39]:
df_user_id_1=df_ords[df_ords['user_id']=='1']

In [40]:
df_user_id_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# Direction 10:

#### You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [41]:
df_user_id_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# Exporting Data Frames

In [42]:
# exporting orders wrangled
df_ords.to_csv(r'C:\Users\Admin\Desktop\Data Tasks\Data Immersion\Achievement 4\Instacart Basket Analysis\02 data\Prepared Data\orders_wrangled.csv')

In [43]:
df_dep_t_new.to_csv(r'C:\Users\Admin\Desktop\Data Tasks\Data Immersion\Achievement 4\Instacart Basket Analysis\02 data\Prepared Data\departments_wrangled.csv')